extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnController.java

```diff
@@ -0,0 +1,289 @@
+                    ctx.fail(generateServerGetAssertion.cause());
+                    return;
+                }
+
+                final JsonObject getAssertion = generateServerGetAssertion.result();
+
+                authMech.getLoginManager().save(getAssertion.getString("challenge"), ctx, CHALLENGE_COOKIE, null,
+                        ctx.request().isSSL());
+                authMech.getLoginManager().save(username, ctx, USERNAME_COOKIE, null, ctx.request().isSSL());
+
+                ok(ctx, getAssertion);
+            });
+        } catch (IllegalArgumentException e) {
+            ctx.fail(400, e);
+        } catch (RuntimeException e) {
+            ctx.fail(e);
+        }
+
+    }
+
+    /**
+     * Endpoint for callback
+     *
+     * @param ctx the current request
+     */
+    public void callback(RoutingContext ctx) {
+        try {
+            // might throw runtime exception if there's no json or is bad formed
+            final JsonObject webauthnResp = ctx.getBodyAsJson();
+            // input validation
+            if (webauthnResp == null ||
+                    !containsRequiredString(webauthnResp, "id") ||
+                    !containsRequiredString(webauthnResp, "rawId") ||
+                    !containsRequiredObject(webauthnResp, "response") ||
+                    !containsOptionalString(webauthnResp.getJsonObject("response"), "userHandle") ||
+                    !containsRequiredString(webauthnResp, "type") ||
+                    !"public-key".equals(webauthnResp.getString("type"))) {
+
+                ctx.fail(400, new IllegalArgumentException(
+                        "Response missing one or more of id/rawId/response[.userHandle]/type fields, or type is not public-key"));
+                return;
+            }
+
+            RestoreResult challenge = authMech.getLoginManager().restore(ctx, CHALLENGE_COOKIE);
+            RestoreResult username = authMech.getLoginManager().restore(ctx, USERNAME_COOKIE);
+            if (challenge == null || challenge.getPrincipal() == null || challenge.getPrincipal().isEmpty()
+                    || username == null || username.getPrincipal() == null || username.getPrincipal().isEmpty()) {
+                ctx.fail(400, new IllegalArgumentException("Missing challenge or username"));
+                return;
+            }
+
+            ManagedContext requestContext = Arc.container().requestContext();
+            requestContext.activate();
+            ContextState contextState = requestContext.getState();
+            // input basic validation is OK
+            // authInfo
+            WebAuthnCredentials credentials = new WebAuthnCredentials()
+                    .setOrigin(origin)
+                    .setDomain(domain)
+                    .setChallenge(challenge.getPrincipal())
+                    .setUsername(username.getPrincipal())
+                    .setWebauthn(webauthnResp);
+            identityProviderManager
```

Comment:
This is code is executed outside of `WebAuthn` `HttpAuthenticationMechanism` - which I think should be the entry point for `WebAuthn` authenticating the user

Reply:
You mean the callbacks are defined as autonomous routes and not as part of the `HttpAuthenticationMechanism` interception like OIDC? I thought that was really weird and confusing to do, when they have their own URI mapping and can't really interfere in other requests, unlike cookie or header-based auth like Basic Auth session cookies.
Is there a real benefit in doing it the complex way? This seems much cleaner.

Reply:
@FroMage What I'm confused about (or concerned is probably the better word) is that we have several paths for obtaining the identity, can you clarify please when is the `WebAuthn` auth mechanism is used and when via these autonomous routes ?
Perhaps it is a neater way to do it, but I'd say we should review how to do it consistently across all of the mechanisms.

Reply:
@FroMage I think I understand it better now after looking through the code again. But the challenge is different here, with `WebAuthnAuthenticationMechanism` the failure results in a redirect to the login page challenge - but with the custom callback it is 401 - is it expected for this use case ?

Reply:
I don't think so: with a custom endpoint I use `WebAuthnSecurity.login/register` and I get a `Uni` with an exception and then it's up to the custom endpoint to figure out what to do.

Reply:
@FroMage It does look like it terminates with 401 here, https://github.com/quarkusio/quarkus/pull/24316/files#diff-171ecf0519a39e39e4ae595be50469f3e93f0caf49494561f0075e7b05330ddeR254 ?

Reply:
Never mind - that is a bad request 400